liberum_vault/src/vault.rs

```diff
@@ -0,0 +1,481 @@
+                    key_u64_slice[2] as i64,
+                    key_u64_slice[3] as i64,
+                ];
+
+                let fragment = stmt
+                    .query_row(key_as_i64, |r| {
+                        let path: String = match r.get(0) {
+                            Ok(p) => p,
+                            Err(e) => return Result::Err(e),
+                        };
+
+                        let size: u64 = match r.get(1) {
+                            Ok(s) => s,
+                            Err(e) => return Result::Err(e),
+                        };
+
+                        Result::Ok(FragmentInfo::new(key, Path::new(&path), size))
+                    })
+                    .optional()?;
+
+                Ok(fragment)
+            })
+            .await
+            .map_err(|e| anyhow!(e))
+    }
+
+    async fn store_fragment_info(&self, fragment: FragmentInfo) -> Result<()> {
+        const SELECT_FRAGMENT_QUERY: &str = "SELECT COUNT(*) FROM fragment WHERE hash0 = ?1 AND hash1 = ?2 AND hash2 = ?3 AND hash3 = ?4";
+        const INSERT_FRAGMENT_QUERY: &str = "
+                INSERT INTO fragment (hash0, hash1, hash2, hash3, path, size)
+                VALUES (?1, ?2, ?3, ?4, ?5, ?6)
+            ";
+
+        let hash_as_u64 = fragment.hash.as_u64_slice_be();
+
+        let cnt = self
+            .db
+            .call(move |conn| {
+                let key_as_i64 = [
+                    hash_as_u64[0] as i64,
+                    hash_as_u64[1] as i64,
+                    hash_as_u64[2] as i64,
+                    hash_as_u64[3] as i64,
+                ];
+
+                let cnt = conn.query_row(SELECT_FRAGMENT_QUERY, key_as_i64, |r| {
+                    let cnt: usize = r.get(0)?;
+
+                    Ok(cnt)
+                })?;
+
+                Ok(cnt)
+            })
+            .await?;
+
+        if cnt != 0 {
+            return Ok(());
+        }
+
+        self.db
+            .call(move |conn| {
+                conn.execute(
+                    INSERT_FRAGMENT_QUERY,
+                    (
+                        hash_as_u64[0] as i64,
+                        hash_as_u64[1] as i64,
+                        hash_as_u64[2] as i64,
+                        hash_as_u64[3] as i64,
+                        fragment.path.to_str(),
+                        fragment.size,
+                    ),
+                )?;
+
+                Ok(())
+            })
+            .await
+            .map_err(|e| anyhow!(e))
+    }
+
+    async fn load_fragment(&self, key: Key) -> Result<Option<FragmentData>> {
+        let fragment_info = self.load_fragment_info(key.clone()).await?;
+
+        if let None = fragment_info {
+            return Ok(None);
+        }
+
+        let fragment_info = fragment_info.unwrap();
+        let fragment_path = fragment_info.path;
+        let fragment_file = File::open(&fragment_path).await?;
+
+        // TODO: Fix, is there any way to not use this take here?
```

Comment:
why not?

Reply:
Because it is redundant, I should remove it, because I think I fixed the problem with stream I had earlier.

Reply:
Done, tests all green so it should be good.

Reply:
I thought I don't understand something, because I tried removing this method call locally and it was working.

Reply:
Yeah, because it works without this take method -- it does nothing, so it may be removed without any effect.